src/site/content/en/blog/serial/index.md

```diff
@@ -0,0 +1,610 @@
+
+When you receive data from the serial device, you won't necessarily get all of
+the data at once. It may be arbitrarily chunked. For more information, see
+[Streams API concepts].
+
+To deal with this, you can use some built-in transform streams such as
+`TextDecoderStream` or create your own transform stream which allows you to
+parse the incoming stream and return parsed data. The transform stream sits
+between the serial device and the read loop that is consuming the stream. It can
+apply an arbitrary transform before the data is consumed. Think of it like an
+assembly line: as a widget comes down the line, each step in the line modifies
+the widget, so that by the time it gets to its final destination, it's a fully
+functioning widget.
+
+<figure class="w-figure">
+  <img src="./aeroplane-factory.jpg" alt="Photo of an aeroplane factory">
+  <figcaption class="w-figcaption">World War 2 Castle Bromwich Aeroplane Factory</figcaption>
+</figure>
+
+For example, consider how to create a transform stream class that consumes a
+stream and chunks it based on line breaks. Its `transform()` method is called
+every time new data is received by the stream. It can either enqueue the data or
+save it for later. The `flush()` method is called when the stream is closed, and
+it handles any data that hasn't been processed yet.
+
+To use the transform stream class, you need to pipe an incoming stream through
+it. In the readableStream example under [Read from a serial port](#read-port),
+the original input stream was only piped through a `TextDecoderStream`, so we
+need to call `pipeThrough()` to pipe it through our new `LineBreakTransformer`.
+
+```js
+class LineBreakTransformer {
+  constructor() {
+    // A container for holding stream data until a new line.
+    this.chunks = "";
+  }
+
+  transform(chunk, controller) {
+    // Append new chunks to existing chunks.
+    this.chunks += chunk;
+    // For each line breaks in chunks, send the parsed lines out.
+    const lines = this.chunks.split("\r\n");
+    this.chunks = lines.pop();
+    lines.forEach((line) => controller.enqueue(line));
+  }
+
+  flush(controller) {
+    // When the stream is closed, flush any remaining chunks out.
+    controller.enqueue(this.chunks);
+  }
+}
+```
+
+```js/2-5
+const textDecoder = new TextDecoderStream();
+const readableStreamClosed = port.readable.pipeTo(textDecoder.writable);
+const reader = textDecoder.readable
+  .pipeThrough(new TransformStream(new LineBreakTransformer()))
+  .getReader();
+```
+
+For debugging serial device communication issues, use the `tee()` method of
+`port.readable` to split the streams going to or from the serial device. The two
+streams created can be consumed independently and this allows you to print one
+to the console for inspection.
+
+```js
+const [appReadable, devReadable] = port.readable.tee();
+
+// TODO: Update UI with incoming data from appReadable.
+// TODO: Log incoming data in JS console for inspection from devReadable.
```

Comment:
I'm thinking "TODO" should be removed. People might think we neglected to finish the code example.

Reply:
Fixed.